### PR TITLE
feat: postcss 8 and new modern postcss API

### PR DESCRIPTION
**Pull request by anikethsaha**
Adding support for new postcss 8 and its new API

#### Status

- [x] dep updates
- [x] test helpers
- [x] site playground
- [x] cssnano core
- [x] postcss-colormin
- [x] postcss-convert-values
- [x] postcss-discard-comments
- [x] postcss-discard-empty
- [x] postcss-discard-duplicates
- [x] postcss-discard-overridden (the test runner needs to be fixed)
- [x] postcss-discard-unused
- [x] postcss-merge-idents
- [x] postcss-merge-longhand
- [x] postcss-merge-rules
  - [x] plugin refactor
  - [x] tests case for `should not crash when node.raws.value is null` not working properly as postcss runner is not able to run the plugin - maybe the postcss is `< 8` there, not sure. error `'true' is not a postcss plugin` when using the `postcss-simple-vars` plugin
- [x] postcss-minify-font-values
- [x] postcss-minify-gradients
- [x] postcss-minify-params
- [x] postcss-minify-selectors
  - [x] the plugin code
  - [x] tests: the `toModules` plugin's visitor keys are not being visited. But the plugin is being called. Need to fix that.
  - [ ] (non-blocker): https://github.com/jonathantneal/postcss-font-magician/issues/91
- [ ] postcss-normalize-charset
- [ ] postcss-normalize-display-values
- [ ] postcss-normalize-positions
- [ ] postcss-normalize-repeat-style
- [ ] postcss-normalize-string
- [ ] postcss-normalize-timing-functions
- [ ] postcss-normalize-unicode
- [ ] postcss-normalize-url
- [ ] postcss-normalize-whitespace
- [ ] postcss-ordered-values
- [ ] postcss-reduced-idents
- [ ] postcss-reduced-initial
- [ ] postcss-reduced-transform
- [ ] postcss-svgo
- [ ] postcss-unique-selectors
- [ ] postcss-zindex
- [ ] stylehacks
- [ ] postcss-calc (external)
  - [ ] it is supported in upstream
  - [ ] need to update dependency in our end
- [ ] css-declaration-sorter (external)
  - [x] it is supported in upstream https://github.com/Siilwyn/css-declaration-sorter/issues/106
  - [ ] need to update dependency in our end
- [ ] autoprefixer (external) 
  - [x] it is supported in upstream
  - [ ] need to update dependency in our end

### CODE REVIEW

File: packages/postcss-minify-selectors/src/__tests__/index.js [hunk 1]
```diff
@@ -505,6 +484,8 @@ test('should handle selectors from other plugins', () => {
   padding: 4px;
 }`;
   expect(
+    // eslint-disable-next-line no-warning-comments
+    // FIXME: not working as `toModules` plugin is not running properly
     postcss([toModules, plugin]).process(css, { from: undefined }).css
```
Comment (anikethsaha) [inline packages/postcss-minify-selectors/src/__tests__/index.js:489]: @ai here, the `toModules` plugin is not working properly as its visitor keys are not being visited. Though the function is being called. Thats why the plugin is not transforming the selectors and the test case is failing

The definition for the plugin is [here](https://github.com/cssnano/cssnano/blob/9cfe8113ce30b4533bfc7be873e5c4055effb110/packages/postcss-minify-selectors/testPlugin/toModules.js)

Any suggestion why it is happening ?

Comment (ai) [reply packages/postcss-minify-selectors/src/__tests__/index.js:489]: Everything looks good.

Can you debug this block of the code? https://github.com/postcss/postcss/blob/master/lib/lazy-result.js#L202-L213

Comment (anikethsaha) [reply packages/postcss-minify-selectors/src/__tests__/index.js:489]: how ? the postcss dep is transpiled so cant really locate this one.

**Review by evilebottnawi — COMMENTED** [review]
Can you separate PRs, one plugin = one PR, PostCSS 8 should have compatibility with existing plugins, so we will have only warnings, if you faced with error, need report to `postcss` repo

**Comment by anikethsaha** [comment]
I guess there would be some issues as we have our own runner. [here](https://github.com/cssnano/cssnano/blob/master/packages/cssnano/src/index.js#L8)

here is the [diff](https://github.com/cssnano/cssnano/pull/954/files#diff-0608d0a2bd2377af41159e54b291a3d6R117-R139)

So I guess it is better to do this in a single one itself. We do have nightly as well and implementing some support for postcss 8 and rest in postcss 7 might break.

File: packages/postcss-minify-selectors/src/__tests__/index.js (hunk 1 above)
Comment (ai) [reply packages/postcss-minify-selectors/src/__tests__/index.js:489]: What is in `postcss` dependency on this line? I think it should be `node_modules/postcss` in your project directory.

Open `node_modules/postcss/lib/lazy-result.js` in a text editor and put `console.log()` or use your favorite debugger. Try to look at what is in `this.listeners` (that PostCSS correctly cached `toModule` listener). Try to look at `this.plugins` to be sure that `toModule` was saved. `this.walkSync` calls will help you to understand what nodes were scanned.

Sorry, you need to help me. I am right now on other PostCSS 8 issue. 50M downloads lead to an insane number of edge cases.

Comment (anikethsaha) [reply packages/postcss-minify-selectors/src/__tests__/index.js:489]: > What is in postcss dependency on this line? I think it should be node_modules/postcss in your project directory.

Yup I am checking this dir only, but the code is transpiled. 

image


is it postcss@7 ?

Comment (ai) [reply packages/postcss-minify-selectors/src/__tests__/index.js:489]: It is PostCSS 7

Comment (anikethsaha) [reply packages/postcss-minify-selectors/src/__tests__/index.js:489]: okay....looking into it.

Comment (anikethsaha) [reply packages/postcss-minify-selectors/src/__tests__/index.js:489]: @ai  

here is the output

image

Also, `toModules` has been registered. 

image

Comment (ai) [reply packages/postcss-minify-selectors/src/__tests__/index.js:489]: It means that an issue somewhere in that code https://github.com/postcss/postcss/blob/master/lib/lazy-result.js#L391-L422

Can you check how the algorithm is working there?

Comment (anikethsaha) [reply packages/postcss-minify-selectors/src/__tests__/index.js:489]: yea, checking....

Comment (anikethsaha) [reply packages/postcss-minify-selectors/src/__tests__/index.js:489]: Okay, I think some issue with normalizing the plugins. 

if the plugin type is function, should it be converted to object or should it be called before pushing to normalized plugins ?

cause [here in prepareVisitor](https://github.com/postcss/postcss/blob/master/lib/lazy-result.js#L392) my plugins are still functioning so the check would not be true and their wont be any listeners saved so the `this.hasListeners` check [here](https://github.com/postcss/postcss/blob/master/lib/lazy-result.js#L350) will fail and the plugin visitor keys wont be call.

I changed [this](https://github.com/postcss/postcss/blob/master/lib/processor.js#L50) to `normalized.push(i())` and now the visitor keys are working (not sure about any side effect for this).

Comment (ai) [reply packages/postcss-minify-selectors/src/__tests__/index.js:489]: Seems like you forget to set `module.exports.postcss = true` and this why this line was not called https://github.com/postcss/postcss/blob/master/lib/processor.js#L40

Comment (anikethsaha) [reply packages/postcss-minify-selectors/src/__tests__/index.js:489]: its there already, right ! .

[`plugin`](https://github.com/cssnano/cssnano/pull/954/files#diff-20446d189f703b4719024d8350448ef9R219)
[`toModules`](https://github.com/cssnano/cssnano/pull/954/files#diff-95973145503262cae31de6e4b30735acR26)

Comment (ai) [reply packages/postcss-minify-selectors/src/__tests__/index.js:489]: Named export do not work like a property. You,  need to set `pluginCreator.postcss = true`, add a property to default exported function

Comment (anikethsaha) [reply packages/postcss-minify-selectors/src/__tests__/index.js:489]: ohh, let me change this then

**Comment by ludofischer** [comment]
Hi, would you find it useful if I converted some of the remaining plugins myself? I could open a pull request against your postcss-8 branch.

**Comment by anikethsaha** [comment]
> Hi, would you find it useful if I converted some of the remaining plugins myself? I could open a pull request against your postcss-8 branch.

Sounds good. @evilebottnawi do we have to wait for the postcss fixes you were working on to continue in this?

**Comment by evilebottnawi** [comment]
@anikethsaha I think all problems fixes in 8.1, so we can continue to update

**Comment by ludofischer** [comment]
Shall I start from `postcss-normalize-charset` and go down the checkbox list?

**Comment by anikethsaha** [comment]
> Shall I start from `postcss-normalize-charset` and go down the checkbox list?

Sure,
I would recommend not to update the integration test's fixtures. We will do that once everything is migrated. You can update rhe plugin's tests.

**Comment by ludofischer** [comment]
All right. BTW, when I run  `yarn test:only` on the `postcss-8` branch,  Jest hangs. It prints the test results summary so it does manages to run all the tests anyway.

**Comment by alexander-akait** [comment]
Done in master

**Comment by frederikhors** [comment]
Please can you release it? The world is waiting for you.

**Comment by alexander-akait** [comment]
It would be nice if the world helped me and not wait... we always glad to see new contributors

**Comment by frederikhors** [comment]
Believe me, I would love to know where to start to help you. Geniuses like you often underestimate themselves.

**Comment by ludofischer** [comment]
I would gladly help tag and publish a release candidate, but I haven't got the permissions though. Don't know if there are changes that still need to be merged.

**Pull request closed without merging**